sympy/solvers/diophantine/tests/test_diophantine.py

```diff
@@ -623,6 +623,17 @@ def test_diop_general_sum_of_squares_quick():
     raises(NotImplementedError, lambda: classify_diop(-eq))
 
 
+def test_issue_23807():
+    # fixes recursion error
+    eq = x**2 + y**2 + z**2 - 1000000
+    base_soln = {(0, 0, 1000), (0, 352, 936), (480, 600, 640), (24, 640, 768), (192, 640, 744),
+                 (192, 480, 856), (168, 224, 960), (0, 600, 800), (280, 576, 768), (152, 480, 864),
+                 (0, 280, 960), (352, 360, 864), (424, 480, 768), (360, 480, 800), (224, 600, 768),
+                 (96, 360, 928), (168, 576, 800), (96, 480, 872)}
+
+    assert diophantine(eq) == base_soln
```

Comment:
Shouldn't this also return negative solutions?

```
In [3]: diophantine(x**2 + y**2 - 10)
Out[3]: {(-3, -1), (-3, 1), (-1, -3), (-1, 3), (1, -3), (1, 3), (3, -1), (3, 1)}

In [4]: diophantine(x**2 + y**2 + z**2 - 10)
Out[4]: {(0, 1, 3)}
```

Reply:
Hmm, that is correct. How are there not any other checks for this, or did other people mess this up too? 😅 I will work on it when I get home.

Reply:
Hey, thank you for the review. As it turns out, whether it should return negative solutions or not isn't really specified in the documentations. In both the previous version (say release 1.11) and this PR I made, the following behaviour is equivalent:

```python
from sympy.abc import a, b, c, d, e, x, y, z
from sympy.solvers.diophantine.diophantine import diophantine

if __name__ == '__main__':
    eq1 = a**2 + b**2 - 234
    eq2 = a**2 + b**2 + c**2 - 234
    eq3 = a**2 + b**2 + c**2 + d**2 + e**2 - 234
    for eq in [eq1, eq2, eq3]:
        sols = diophantine(eq)
        assert len(sols) > 0
        print([sol for sol in sols if any(x < 0 for x in sol)])

"""
[(-3, -15), (3, -15), (-15, -3), (-15, 3), (15, -3), (-3, 15)]
[]
[]
"""
```

Which means it only return all signs (and permutations, which seems to be unintended... otherwise there won't be a special `permute=True` flag for it) for the case $x^2+y^2=N$.

I believe it is better to start another PR/issue to fix this by unifying the behaviours of the 8 solvers used by `diop_solve`.

Reply:
> better to start another PR/issue to fix this by unifying the behaviours of the 8 solvers

Sounds good.

Reply:
It looks like the permute argument is supposed to control this:
```python
In [6]: diophantine(x**2 + y**2 + z**2 - 10)
Out[6]: {(0, 1, 3)}

In [7]: diophantine(x**2 + y**2 + z**2 - 10, permute=True)
Out[7]: 
{(-3, -1, 0), (-3, 0, -1), (-3, 0, 1), (-3, 1, 0), (-1, -3, 0), (-1, 0, -3), (-1, 0, 3), (-1, 3, 0)
, (0, -3, -1), (0, -3, 1), (0, -1, -3), (0, -1, 3), (0, 1, -3), (0, 1, 3), (0, 3, -1), (0, 3, 1), (
1, -3, 0), (1, 0, -3), (1, 0, 3), (1, 3, 0), (3, -1, 0), (3, 0, -1), (3, 0, 1), (3, 1, 0)}
```

Reply:
@oscarbenjamin Yes for sum of more than two squares `permute=True` controls the signs and permutations of the output, I was explaining why for two squares i.e. `diophantine(x**2 + y**2 - 13)` it behaves like `permute=True`, even though it's False by default.

Reply:
Oh I see you already mentioned permute.

Yes, the docstring does need to be improved. A note could be added or perhaps just the linear "by default only the base solution is returned" could be changed.

The start of the docstring really needs rewriting as well:
https://github.com/sympy/sympy/blob/78cc8127da4f3f281222e9a7bca555bf30979bac/sympy/solvers/diophantine/diophantine.py#L1273-L1277
That should say something like "Solve diophantine equations" or "Find integer solutions to polynomial equations" or something.

Also the whole interface should just be rewritten. It's a terrible interface (#20682).